query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+        }
+
+        let mut args = where_fn.args.into_iter();
+
+        // TODO: process source variables.
+        match args.next().unwrap() {
+            FnArg::SrcVar(SrcVar::DefaultSrc) => {},
+            _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "source variable".into(), 0)),
+        }
+
+        if where_fn.binding.is_empty() {
+            // The binding must introduce at least one bound variable.
+            bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::NoBoundVariable));
+        }
+
+        if !where_fn.binding.is_valid() {
+            // The binding must not duplicate bound variables.
+            bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::RepeatedBoundVariable));
+        }
+
+        // Turn input constant into a matrix, ready to feed into a VALUES clause.
+        let constant = args.next().unwrap();
+        let rows = { // Scope borrow of self.
+            let operator = where_fn.operator;
+            let lookup = |in_var| {
+                if !self.input_variables.contains(&in_var) {
+                    bail!(ErrorKind::UnboundVariable((*in_var.0).clone()));
+                }
+                match self.bound_value(&in_var) {
+                    Some(ref in_value) => Ok(in_value.clone()),
+                    None => bail!(ErrorKind::UnboundVariable((*in_var.0).clone())),
+                }
+            };
+
+            match where_fn.binding {
+                Binding::BindScalar(_) => constant.into_unit_matrix(&lookup)?,
+                Binding::BindTuple(_) => constant.into_row_matrix(&lookup)?,
+                Binding::BindColl(_) => constant.into_column_matrix(&lookup)?,
+                Binding::BindRel(_) => constant.into_block_matrix(&lookup)?,
+            }
+        };
+
+        // Typecheck and drop columns that are not actually bound.  By restricting to homogeneous
+        // columns, we greatly simplify projection.  In the future, we could loosen this
+        // restriction, at the cost of projecting (some) value type tags.  If and when we want to
+        // algebrize in two phases and allow for late-binding input variables, we'll probably be
+        // able to loosen this restriction with little penalty.
+        let variables: Vec<Option<Variable>> = where_fn.binding.variables();
+        let (values, column_types) = reduced_rows_and_column_types(rows, &variables.iter().map(|x| x.is_some()).collect())?;
+        let names: Vec<Variable> = variables.into_iter().filter_map(|x| x).collect();
+
+        let named_values = ComputedTable::NamedValues {
+            names: names.clone(),
+            values: values,
+        };
+        let table = self.computed_tables.push_computed(named_values);
+        let alias = self.next_alias_for_table(table);
+
+        // Constrain the types of the bound variables we saw.
```

Comment:
For scalar and tuple binding forms, I think we should statically bind the value in the CC right here.

Furthermore, I think you should check to see whether any of the variables have already been used in the query. If not, there's no point in producing SQL at all; any use of the grounded variable (either in projection or in a pattern) will pick up the constant from the CC. Treat `ground` very much like `:in` at that point.

That makes a query like

```
[:find ?x ?v :where
 [(ground 10) ?v]
 [?x :foo/v ?v]]
```

very efficient… and you might think that's silly, but it's an approach used for branching, even with `:in`:

```
[:find ?x ?v
 :in ?v1 ?v2
 :where
 (or (and [(ground ?v1) ?v]
          [?x :foo/v ?v])
    …)]
```

I think it's also worth thinking about different representations for each of the binding lists.

A scalar can be expressed as a very efficient subquery, even if you don't manage to substitute it, and you can also project the type code at this point.

```
sqlite> select x from (SELECT 5 AS x);
x
----------
5
```